crates/punktf-lib/src/visit/diff/mod.rs

```diff
@@ -0,0 +1,210 @@
+//! A [`Visitor`](`crate::visit::Visitor`) implementation which creates events for
+//! files which differ from the content it would have once deployed.
+
+use crate::{
+	profile::LayeredProfile,
+	profile::{source::PunktfSource, transform::Transform},
+	visit::*,
+};
+use std::path::Path;
+
+/// Applies any relevant [`Transform`](`crate::profile::transform::Transform`)
+/// for the given file.
+fn transform_content(profile: &LayeredProfile, file: &File<'_>, content: String) -> String {
+	let mut content = content;
+
+	// Copy so we exec_dotfile is not referenced by this in case an error occurs.
+	let exec_transformers: Vec<_> = file.dotfile().transformers.to_vec();
+
+	// Apply transformers.
+	// Order:
+	//   - Transformers which are specified in the profile root
+	//   - Transformers which are specified on a specific dotfile of a profile
+	for transformer in profile.transformers().chain(exec_transformers.iter()) {
+		content = transformer.transform(content).unwrap();
+	}
+
+	content
+}
+
+/// An event which is emitted for every differing item.
+#[derive(Debug)]
+pub enum Event<'a> {
+	/// File does currently not exist but would be created.
+	NewFile(&'a Path),
+
+	/// Directory does currently not exist but would be created.
+	NewDirectory(&'a Path),
+
+	/// File does exist but the contents would changed.
+	Diff {
+		/// Absoulte path to the target location.
+		target_path: &'a Path,
+
+		/// Contents of the current file on the filesystem.
+		old_content: String,
+
+		/// Contents of the file after a deployment.
+		new_contnet: String,
+	},
+}
+
+impl Event<'_> {
+	/// Returns the absolute target path for the diff.
+	pub const fn target_path(&self) -> &Path {
+		match self {
+			Self::NewFile(p) => p,
+			Self::NewDirectory(p) => p,
+			Self::Diff { target_path, .. } => target_path,
+		}
+	}
+}
+
+/// A [`Visitor`](`crate::visit::Visitor`) implementation which checks for
+/// changes which would be made by a deployment.
+/// For each change an [`Event`] is emitted which can be processed by [`Diff.0`].
+#[derive(Debug, Clone, Copy)]
+pub struct Diff<F>(F);
+
+impl<F> Diff<F>
+where
+	F: Fn(Event<'_>),
+{
+	/// Creates a new instance of the visitor.
+	pub const fn new(f: F) -> Self {
+		Self(f)
+	}
+
+	/// Runs the visitor to completion for a given profile.
+	pub fn diff(self, source: &PunktfSource, profile: &mut LayeredProfile) {
+		let mut resolver = ResolvingVisitor(self);
+		let walker = Walker::new(profile);
+		walker.walk(source, &mut resolver).unwrap();
+	}
+
+	/// Emits the given event.
+	fn dispatch(&self, event: Event<'_>) {
+		(self.0)(event)
+	}
+}
+
+impl<F> Visitor for Diff<F>
```

Comment:
i like the visitor pattern, but maybe add some docs

Reply:
Will do. Thanks!